src/compiler/checker.ts

```diff
@@ -23007,13 +23007,20 @@ namespace ts {
 
                 case SyntaxKind.NumericLiteral:
                     // index access
-                    if (node.parent.kind === SyntaxKind.ElementAccessExpression && (<ElementAccessExpression>node.parent).argumentExpression === node) {
-                        const objectType = getTypeOfExpression((<ElementAccessExpression>node.parent).expression);
-                        return getPropertyOfType(objectType, (<NumericLiteral>node).text as __String);
-                    }
-                    break;
+                    const objectType = (() => {
+                        switch (node.parent.kind) {
+                            case SyntaxKind.ElementAccessExpression:
+                                const { argumentExpression, expression } = node.parent as ElementAccessExpression;
+                                return argumentExpression === node ? getTypeOfExpression(expression) : undefined;
+                            case SyntaxKind.LiteralType:
+                                return isIndexedAccessTypeNode(node.parent.parent) ? getTypeFromTypeNode(node.parent.parent.objectType) : undefined;
+                        }
+                    })();
+                    return objectType && getPropertyOfType(objectType, escapeLeadingUnderscores((node as StringLiteral | NumericLiteral).text));

```

Comment:
Can you construct a test case where the lack of escapeLeadingUnderscores makes a string literal element access fail? It would be something like x["__foo"], right?

Reply:
Turns out I need to update a few baselines related to this.